File: apps/web/src/lib/study-guide.ts
import { api } from "./api";

export interface StudyExercise {
  id: string;
  question: string;
  hint: string;
}

export interface StudySection {
  id: string;
  title: string;
  /** Markdown notes. */
  summary: string;
  key_points: string[];
  exercises: StudyExercise[];
}

export interface StudyGuide {
  id: string;
  book_id: string;
  sections: StudySection[];
  created_at: string;
}

export interface GradeResult {
  exercise_id: string;
  correct: boolean;
  feedback: string;
  model_answer: string;
}

/** Throws ApiError(404) when no guide has been generated yet. */
export async function getStudyGuide(bookId: string): Promise<StudyGuide> {
  return api<StudyGuide>(`/books/${bookId}/study-guide`);
}

export async function generateStudyGuide(bookId: string): Promise<StudyGuide> {
  return api<StudyGuide>(`/books/${bookId}/study-guide`, {
    method: "POST",
    body: {},
  });
}

export async function gradeExercise(
  bookId: string,
  exerciseId: string,
  answer: string,
): Promise<GradeResult> {
  return api<GradeResult>(`/books/${bookId}/study-guide/grade`, {
    method: "POST",
    body: { exercise_id: exerciseId, answer },
  });
}

File: apps/api/app/models/study_guide.py
"""StudyGuide — AI-generated study material for a book/video.

One guide per (user, book). The ``sections`` JSONB holds the full structure
(including each exercise's reference answer, which is kept server-side and
never sent to the client) so we can grade free-text answers later without a
second generation pass.
"""
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base


class StudyGuide(Base):
    __tablename__ = "study_guides"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    book_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("books.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )

    # Sections: list of
    #   {id, title, summary (markdown), key_points: [str],
    #    exercises: [{id, question, reference_answer, hint}]}
    # reference_answer stays server-side — used only when grading.
    sections: Mapped[list] = mapped_column(JSONB, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    __table_args__ = (
        UniqueConstraint("user_id", "book_id", name="uq_study_guide_user_book"),
    )
